### PR TITLE
Hearts

### PR DESCRIPTION
**Pull request by hopshackle**
Implementation of Hearts

### CODE REVIEW

File: src/main/java/games/hearts/HeartsForwardModel.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        }
+
+        hgs.playerDecks = new ArrayList<>();
+        hgs.drawDeck = FrenchCard.generateDeck("DrawDeck", CoreConstants.VisibilityMode.HIDDEN_TO_ALL);
+        hgs.drawDeck.shuffle(new Random(System.currentTimeMillis()));
+
+        int numOfPlayers = hgs.getNPlayers();
+
+        Map<Integer, List<FrenchCard>> cardsToRemove = new HashMap<>();
+        cardsToRemove.put(3, Arrays.asList(new FrenchCard(FrenchCard.FrenchCardType.Number, FrenchCard.Suite.Diamonds, 2)));
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsForwardModel.java:61]: Lots of hardcoded numbers.

File: src/main/java/games/hearts/HeartsForwardModel.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            int numberOfCards = getNumberOfCards(hgs);
+
+            for (int card = 0; card < numberOfCards; card++) {
+                playerDeck.add(hgs.drawDeck.draw());
+            }
+        }
+    }
+
+    private static int getNumberOfCards(HeartsGameState hgs) {
+        int numberOfCards;
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsForwardModel.java:100]: Parameter

File: src/main/java/games/hearts/HeartsForwardModel.java (hunk 1 above)
Comment (rdgain) [reply src/main/java/games/hearts/HeartsForwardModel.java:61]: Everywhere!

File: src/main/java/games/hearts/HeartsGameState.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+    @Override
+    protected GameType _getGameType() {
+        return GameType.Hearts;
+    }
+
+    @Override
+    protected List<Component> _getAllComponents() {
+        return new ArrayList<Component>() {{
+            addAll(playerDecks);
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsGameState.java:52]: Not all are listed?

File: src/main/java/games/hearts/HeartsGameState.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            if (trickDeck != null) {
+                int points = 0;
+
+                // Iterate over all cards in the trick deck
+                for (FrenchCard card : trickDeck.getComponents()) {
+                    if (card.suite == FrenchCard.Suite.Hearts) {
+                        points += params.heartCard;
+                    }
+                    // The queen of spades is worth 13 points
+                    else if (card.suite == FrenchCard.Suite.Spades && card.number == 12) {
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsGameState.java:85]: Hard-coded number?

**Review by rdgain — CHANGES_REQUESTED** [review]

**Comment by hopshackle** [comment]
Various things now parameterised, and getAllComponents now does.

File: src/main/java/games/hearts/HeartsForwardModel.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    public void _afterAction(AbstractGameState gameState, AbstractAction action) {
+        HeartsGameState hgs = (HeartsGameState) gameState;
+
+        if (hgs.getGamePhase() == HeartsGameState.Phase.PASSING) {
+            if (action instanceof Pass) {
+
+                // Check if current player has passed 3 cards
+                int cardsPassed = hgs.pendingPasses.get(hgs.getCurrentPlayer()).size();
+                if (cardsPassed == 3) {
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsForwardModel.java:83]: One more parameter

File: src/main/java/games/hearts/HeartsForwardModel.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+            hgs.playerTricksTaken[winningPlayerID]++;
+
+            hgs.setFirstPlayer(winningPlayerID);
+        }
+        hgs.currentPlayedCards.clear();
+
+        // Check if all cards from player hands have been played
+        if (hgs.playerDecks.stream().allMatch(deck -> deck.getSize() == 0)) {
+            hgs.scorePointsAtEndOfRound();
+            boolean scoreAbove100 = hgs.playerPoints.values().stream().anyMatch(score -> score >= 100);
```
Comment (rdgain) [inline src/main/java/games/hearts/HeartsForwardModel.java:228]: And this one too.

File: src/main/java/games/hearts/HeartsForwardModel.java (hunk 5 above)
Comment (hopshackle) [reply src/main/java/games/hearts/HeartsForwardModel.java:83]: Damn - missed that.

File: src/main/java/games/hearts/HeartsForwardModel.java (hunk 6 above)
Comment (hopshackle) [reply src/main/java/games/hearts/HeartsForwardModel.java:228]: and that one!

**Review by rdgain — APPROVED** [review]

**Pull request merged**